src/core/src/package_managers/AptitudePackageManager.py

```diff
@@ -79,9 +80,55 @@ def __init__(self, env_layer, execution_config, composite_logger, telemetry_writ
+        # type: (str, str) -> str
+        """ Prepares the custom sources list for use in a command. Idempotent. """
+        try:
+            formula = "F-[{0}]-[{1}]".format(max_patch_published_date, base_classification)
+            if self.cached_customer_source_list_formula == formula:
+                return self.custom_sources_list
+            self.cached_customer_source_list_formula = formula
+
+            command = self.cmd_prep_custom_sources_list_template.replace("<CLASSIFICATION>", base_classification if base_classification == "security" else "\"\"")
+            code, out = self.env_layer.run_command_output(command, False, False)
+            sources_content = self.env_layer.file_system.read_with_retry(self.custom_sources_list)
+            self.composite_logger.log_debug("[APM] Modified custom sources list with classification. [Code={0}][Out={1}][Content={2}]".format(str(code), str(out), str(sources_content)))   # non-zero error code to be investigated
+
+            if max_patch_published_date != str():
+                target = "://snapshot.ubuntu.com/ubuntu/{0}".format(self.max_patch_publish_date)
+
+                if "://snapshot.ubuntu.com/ubuntu/" in sources_content:
+                    sources_content_split = sources_content.split(" ")
+                    for i in range(0, len(sources_content_split)):
+                        if "://snapshot.ubuntu.com/ubuntu/" in sources_content_split[i]:
+                            sources_content.replace(sources_content_split[i], target)
+
+                sources_content = sources_content.replace("://azure.archive.ubuntu.com/ubuntu/", target)
+                sources_content = sources_content.replace("://in.archive.ubuntu.com/ubuntu/", target)
+                sources_content = sources_content.replace("://security.ubuntu.com/ubuntu/", target)
+                sources_content = sources_content.replace("http://snapshot.ubuntu.com/", "https://snapshot.ubuntu.com/")
+                self.composite_logger.log_debug("[APM] Modified custom sources list with snapshot. [Code={0}][Out={1}][Content={2}]".format(str(code), str(out), str(sources_content)))
+
+                self.env_layer.file_system.write_with_retry_using_temp_file(self.custom_sources_list, sources_content)
+
+            self.refresh_repo(sources=self.custom_sources_list)
+        except Exception as error:
+            self.composite_logger.log_error("[APM] Error in modifying custom sources list. [Error={0}]".format(repr(error)))
+            return str()    # defaults code to safety
+
+        return self.custom_sources_list
+
+    def refresh_repo(self, sources=str()):
+        self.composite_logger.log("[APM] Refreshing local repo... [Sources={0}]".format(sources if sources != "" else "Default"))
+        self.invoke_package_manager(self.__generate_command(self.cmd_repo_refresh_template, sources))
+
+    @staticmethod
+    def __generate_command(command_template, new_sources_list=str()):
```

Comment:
This function can be generalized to accept the replaceable term as an input (eg: <SOURCES> in this case) and generate a command using the term and template from it's inputs

Can be re-used on lines 359, 380, etc where we do this:
cmd = self.cmd_single_package_check_versions_template.replace('<PACKAGE-NAME>', package_name)

Reply:
PACKAGE-NAME is in angular brackets in that line, github is blocking it

Reply:
Marking as overall improvement in next PR (first comment). I was going to ask what the second comment was but realized it was the text in the first comment you were talking about. But sounds good?

Reply:
Yes sounds good.